src/plugins/discourse/mirror.js

```diff
@@ -343,5 +369,62 @@ export class Mirror implements DiscourseData {
         addPost(post);
       }
     }
+
+    // I don't want to hard code the expected page size, in case it changes upstream.
+    // However, it's helpful to have a good guess of what the page size is, because if we
+    // get a result which is shorter than the page size, we know we've hit the end of the
+    // user's history, so we don't need to query any more.
+    // So, we guess that the largest page size we've seen thus far is likely the page size,
+    // and if we see any shorter pages, we know we are done for that particular user.
+    // If we are wrong about the page size, the worst case is that we do an unnecessary
+    // query when we are actually already done with the user.
+    let possiblePageSize = 0;
+    // TODO(perf): In the best case (there are no new likes), this requires
+    // doing one query for every user who ever commented in the instance. This
+    // is a bit excessive. For each user, we could store when we last checked
+    // their likes, and when they last posted. Then we could only scan users
+    // who we either haven't scanned in the last week, or who have been active
+    // since our last scan. This would likely improve the performance of this
+    // section of the update significantly.
+    for (const user of this.users()) {
+      let offset = 0;
+      let upToDate = false;
+      while (!upToDate) {
+        const likeActions = await this._fetcher.likesByUser(user, offset);
+        possiblePageSize = Math.max(likeActions.length, possiblePageSize);
+        for (const {timestampMs, postId, username} of likeActions) {
+          const alreadySeenThisLike = db
```

Comment:
@wchargin: let me know if there's a way to condense these queries (e.g. to get a "return value" from the INSERT that tells us if the field was already present).

Reply:
There is, and it’s all in the types:
https://github.com/sourcecred/sourcecred/blob/bf5c3a953bfba08501a5e88864d780b4f07bf6c6/flow-typed/npm/better-sqlite3_vx.x.x.js#L87-L88

I’ll do you one better and turn 2*n* queries into 1: #1301, #1302.